### PR TITLE
fix(path-encoding): support punctuation

### PR DESCRIPTION
**Pull request by rudyardrichter**
### Bug Fixes

- Allow resource paths to contain punctuation.

### CODE REVIEW

**Comment by Avantol13** [comment]
The integration tests are failing :cry: something to do with resource creation from usersync it seems like

File: arborist/resource.go [hunk 1]
```diff
@@ -30,6 +31,43 @@ type ResourceOut struct {
...
+
+func UnderscoreEncode(decoded string) string {
+	// Per https://www.ietf.org/rfc/rfc3986.txt, there are the following
+	// unreserved characters (aside from alphanumeric and underscore) which do
+	// not urlencode:
+	//   - `-`
+	//   - `.`
+	//   - `~`
+	// Percent encoding uses hexadecimal, which we mustn't conflict with. So we
+	// use some made-up codes for these which will not overlap.
```
Comment (paulineribeyre) [inline arborist/resource.go]: what happens if you encode _ - . ~ using the percent hexadecimal encoding?

File: arborist/server.go [hunk 2]
```diff
@@ -67,7 +67,7 @@ func (server *Server) Init() (*Server, error) {
 //    `{resourcePath:/[a-zA-Z0-9_\-\/]+}`
```
Comment (paulineribeyre) [inline arborist/server.go]: you can update this comment

File: arborist/resource_test.go [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+	"github.com/stretchr/testify/assert"
+)
+
+func TestEncodeDecode(t *testing.T) {
+	inputs := []string{
+		"test",
+		"!@#$%^&*()`\\=+-_'\"<>?,.",
+		"___ooOO00O___",
+		"foo-=-bar-=-baz",
+		"🙃",
```
Comment (paulineribeyre) [inline arborist/resource_test.go]: it would be nice to also test a string that contains `_S0`

File: arborist/server_test.go [hunk 4]
```diff
@@ -656,6 +638,17 @@ func TestServer(t *testing.T) {
 			assert.NotEqual(t, "", result.Resource.Tag, msg)
 			resourceTag = result.Resource.Tag
 
+			t.Run("Punctuation", func(t *testing.T) {
+				w := httptest.NewRecorder()
+				// missing required field
```
Comment (paulineribeyre) [inline arborist/server_test.go]: go home random comment, you're drunk

File: arborist/resource.go (hunk 1 above)
Comment (rudyardrichter) [reply arborist/resource.go]: That's what [these](https://github.com/uc-cdis/arborist/blob/558f563d7de8d8233f15f6c991c1ea5ec18a9973/arborist/resource.go#L47-L50) are for.
```
fmt.Println(UnderscoreEncode("_-.~"))

_S0_S1_S2_S3
```

File: arborist/resource_test.go (hunk 3 above)
Comment (rudyardrichter) [reply arborist/resource_test.go]: Added in 5899123

File: arborist/server_test.go (hunk 4 above)
Comment (rudyardrichter) [reply arborist/server_test.go]: It has been sent home.

File: arborist/resource.go (hunk 1 above)
Comment (paulineribeyre) [reply arborist/resource.go]: i mean, why `hexadecimal, which we mustn't conflict with`? why not use `_7E` for `~` instead of `_S3`?

Comment (rudyardrichter) [reply arborist/resource.go]: I suppose you could—I might not have a complete understanding—according to [RFC 3986](https://tools.ietf.org/html/rfc3986#section-2.3) those characters specifically are "unreserved", meaning they wouldn't need to be encoded ordinarily (they aren't affected by `url.QueryEscape` for instance). So I wasn't sure about the treatment for those in encoding/decoding in general. You could use `_7E` instead, I think, I just picked something completely different to guarantee it wouldn't collide.

**Pull request merged**